ci/run/bazel_sharding/bazel_sharding.py

```diff
@@ -0,0 +1,434 @@
+    """Extract BazelRules from the XML obtained from ``bazel query --output=xml``."""
+    xml_rules = element.findall("rule")
+    return [BazelRule.from_xml_element(element) for element in xml_rules]
+
+
+def group_rules_by_time_needed(
+    rules: List[BazelRule],
+) -> List[Tuple[float, List[BazelRule]]]:
+    """
+    Return a list of tuples of (timeout in seconds, list of rules)
+    sorted descending.
+    """
+    grouped_rules = defaultdict(list)
+    for rule in rules:
+        grouped_rules[rule.actual_timeout_s].append(rule)
+    for timeout in grouped_rules:
+        grouped_rules[timeout] = sorted(grouped_rules[timeout])
+    return sorted(grouped_rules.items(), key=lambda x: x[0], reverse=True)
+
+
+def get_rules_for_shard_naive(
+    rules_grouped_by_time: List[Tuple[float, List[BazelRule]]], index: int, count: int
+) -> List[str]:
+    """Create shards by assigning the same number of rules to each shard."""
+    all_rules = []
+    for _, rules in rules_grouped_by_time:
+        all_rules.extend(rules)
+    shard = sorted(all_rules)[index::count]
+    return [rule.name for rule in shard]
+
+
+def get_rules_for_shard_optimal(
+    rules_grouped_by_time: List[Tuple[float, List[BazelRule]]], index: int, count: int
+) -> List[str]:
+    """Creates shards by trying to make sure each shard takes around the same time.
+
+    We use a simple heuristic here (as this problem is NP-complete):
+    1. Determine how long one shard would take if they were ideally balanced
+       (this may be impossible to attain, but that's fine).
+    2. Allocate the next biggest item into the first shard that is below the optimum.
+    3. If there's no shard below optimium, choose the shard closest to optimum.
+
+    This works very well for our usecase and is fully deterministic.
+
+    ``rules_grouped_by_time`` is expected to be a list of tuples of
+    (timeout in seconds, list of rules).
+    """
+    # For sanity checks later.
+    all_rules = []
+    for _, rules in rules_grouped_by_time:
+        all_rules.extend(rules)
+
+    # Instantiate the shards, each represented by a list.
+    shards: List[List[BazelRule]] = [list() for _ in range(count)]
+
+    # The theoretical optimum we are aiming for. Note that this may be unattainable
+    # as it doesn't take into account that tests are discrete and cannot be split.
+    # This is however fine, because it should only serve as a guide which shard to
+    # add the next test to.
+    optimum = (
+        sum(timeout * len(rules) for timeout, rules in rules_grouped_by_time) / count
+    )
+
+    def get_next_longest_rule() -> BazelRule:
+        """
+        Get the next longest (taking up the most time) BazelRule from the
+        ``rules_grouped_by_time`` list.
+        """
+        item = None
+        for _, items in rules_grouped_by_time:
+            if items:
+                item = items.pop()
+                break
+        return item
+
+    def add_rule_to_best_shard(rule_to_add: BazelRule):
```

Comment:
This looks good to me, but maybe we can move this out of the local scope and test it in the unit tests